pkgs/test/by-name/check-references/src/main.rs

```diff
@@ -0,0 +1,160 @@
+use rnix::Root;
+use rnix::SyntaxKind::NODE_PATH;
+use std::collections::HashSet;
+use std::env;
+use std::ffi::OsStr;
+use std::fs::read_to_string;
+use std::path::PathBuf;
+mod line_index;
+use line_index::LineIndex;
+use std::io::Result;
+use std::path::Component;
+
+fn resolve_reference(
+    source: &PathBuf,
+    line: usize,
+    reference: PathBuf,
+    pkg_dir: &PathBuf,
+) -> Option<PathBuf> {
+    // The source path is always a Nix file (the one containing the reference), so the unwrap will never fail
+    let mut rel_to_root = source.parent().unwrap().to_path_buf();
+    let rel_source = source.strip_prefix(&pkg_dir).unwrap();
+    for component in reference.components() {
+        match component {
+            Component::CurDir => {}
+            Component::ParentDir => {
+                rel_to_root = match rel_to_root.parent() {
+                    Some(parent) if parent.starts_with(pkg_dir) => parent.to_path_buf(),
+                    _ => {
+                        println!(
+                            "- {}: File {} at line {line} contains the path expression \"{}\" which may point outside the directory of that package.",
+                            pkg_dir.display(),
+                            rel_source.display(),
+                            reference.display());
+                        return None;
+                    }
+                };
+            }
+            Component::Normal(segment) => {
+                rel_to_root = rel_to_root.join(segment);
+                if !rel_to_root.exists() {
+                    return None;
+                }
+            }
+            Component::RootDir | Component::Prefix(_) => {
+                println!(
+                    "- {}: File {} at line {line} contains the path expression \"{}\" which may point outside the directory of that package.",
+                    pkg_dir.display(),
+                    rel_source.display(),
+                    reference.display()
+                );
+                return None;
+            }
+        }
+    }
+
+    Some(rel_to_root)
+}
+
+fn check_path(
+    pkg_dir: &PathBuf,
+    file: &PathBuf,
+    already_checked: &mut HashSet<PathBuf>,
+) -> Result<()> {
+    if !already_checked.insert(file.canonicalize()?) {
+        //println!("Skipping check for file {:?} because it was already checked before", file);
+        return Ok(());
+    }
+
+    if file.is_dir() {
+        for entry in file.read_dir()? {
+            check_path(&pkg_dir, &entry?.path(), already_checked)?
+        }
+        return Ok(());
+    }
+
+    if file.extension() != Some(&OsStr::new("nix")) {
+        return Ok(());
+    }
+
+    let rel_file = file.strip_prefix(&pkg_dir).unwrap();
```

Comment:
I tried running this tool on `pkgs/development/compilers/gcc/` at current nixpkgs `master`, commit e9b86d7b8b0b68bcb4dcf6770ed2e85652b0594e and got this error:

```
$ RUST_BACKTRACE=1 ~/result/bin/check-references pkgs/development/compilers/gcc/
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: StripPrefixError(())', src/main.rs:80:48
stack backtrace:
   0: rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::result::unwrap_failed
   3: check_references::check_path
   4: check_references::check_path
   5: check_references::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

Reply:
Should not be a problem anymore in the [split-off tooling PR](https://github.com/NixOS/nixpkgs/pull/250885).